# 🌦️ Weather Data Analysis
**Analyst: Hannah Oluwajobi**  
**Dataset:** Weather Data (8,784 rows × 8 columns)  
**Tools:** Python, Pandas

This project explores a real-world weather dataset using pandas. Each question demonstrates a different data analysis technique — filtering, groupby, aggregations, string matching, and more.

In [2]:
import pandas as pd

# Load the dataset
df = pd.read_csv('/Users/hannaholuwajobi/Downloads/1. Weather Data.csv')

# Quick overview
print('Shape:', df.shape)
df.head()

Shape: (8784, 8)


,Date/Time,Temp_C,Dew Point Temp_C,Rel Hum_%,Wind Speed_km/h,Visibility_km,Press_kPa,Weather
0,1/1/2012 0:00,-1.8,-3.9,86,4,8.0,101.24,Fog
1,1/1/2012 1:00,-1.8,-3.7,87,4,8.0,101.24,Fog
2,1/1/2012 2:00,-1.8,-3.4,89,7,4.0,101.26,"Freezing Drizzle,Fog"
3,1/1/2012 3:00,-1.5,-3.2,88,6,4.0,101.27,"Freezing Drizzle,Fog"
4,1/1/2012 4:00,-1.5,-3.3,88,7,4.8,101.23,Fog


---
## Q1. Find all the unique 'Wind Speed' values in the data

In [3]:
df['Wind Speed_km/h'].unique()

array([ 4,  7,  6,  9, 15, 13, 20, 22, 19, 24, 30, 35, 39, 32, 33, 26, 44,
       43, 48, 37, 28, 17, 11,  0, 83, 70, 57, 46, 41, 52, 50, 63, 54,  2])

---
## Q2. Find the number of times when the Weather is exactly 'Clear'

In [4]:
# Method 1 - value_counts()
df['Weather'].value_counts()['Clear']

np.int64(1326)

In [5]:
# Method 2 - Filtering
df[df['Weather'] == 'Clear'].shape[0]

1326

In [6]:
# Method 3 - groupby()
df.groupby('Weather').get_group('Clear').shape[0]

1326

---
## Q3. Find the number of times when the Wind Speed was exactly 4 km/h

In [7]:
df[df['Wind Speed_km/h'] == 4].shape[0]

474

---
## Q4. Find out all the Null Values in the data

In [8]:
df.isnull().sum()

Date/Time           0
Temp_C              0
Dew Point Temp_C    0
Rel Hum_%           0
Wind Speed_km/h     0
Visibility_km       0
Press_kPa           0
Weather             0
dtype: int64

✅ **Finding:** No null values in this dataset — it's clean and complete.

---
## Q5. Rename the column 'Weather' to 'Weather Condition'

In [9]:
df.rename(columns={'Weather': 'Weather Condition'}, inplace=True)
df.columns

Index(['Date/Time', 'Temp_C', 'Dew Point Temp_C', 'Rel Hum_%',
       'Wind Speed_km/h', 'Visibility_km', 'Press_kPa', 'Weather Condition'],
      dtype='object')

---
## Q6. What is the mean of each column against each Weather Condition?

In [10]:
df.groupby('Weather Condition').mean(numeric_only=True).round(2)

,Temp_C,Dew Point Temp_C,Rel Hum_%,Wind Speed_km/h,Visibility_km,Press_kPa
Weather Condition,,,,,,
Clear,6.83,0.09,64.50,10.56,30.15,101.59
Cloudy,7.97,2.38,69.59,16.13,26.63,100.91
Drizzle,7.35,5.50,88.24,16.10,17.93,100.44
"Drizzle,Fog",8.07,7.03,93.28,11.86,5.26,100.79
"Drizzle,Ice Pellets,Fog",0.40,-0.70,92.00,20.00,4.00,100.79
"Drizzle,Snow",1.05,0.15,93.50,14.00,10.50,100.89
"Drizzle,Snow,Fog",0.69,0.12,95.87,15.53,5.51,99.28
Fog,4.30,3.16,92.29,7.95,6.25,101.18
Freezing Drizzle,-5.66,-8.00,83.57,16.57,9.20,100.20


---
## Q7. What is the Standard Deviation of 'Pressure' in this data?

In [11]:
df['Press_kPa'].std()

0.8440047459486474

---
## Q8. What is the Variance of 'Relative Humidity' in this data?

In [12]:
df['Rel Hum_%'].var()

286.2485501984998

---
## Q9. Find all instances when 'Snow' was recorded

In [13]:
# Method 1 - value_counts()
df['Weather Condition'].value_counts()

Weather Condition
Mainly Clear                               2106
Mostly Cloudy                              2069
Cloudy                                     1728
Clear                                      1326
Snow                                        390
Rain                                        306
Rain Showers                                188
Fog                                         150
Rain,Fog                                    116
Drizzle,Fog                                  80
Snow Showers                                 60
Drizzle                                      41
Snow,Fog                                     37
Snow,Blowing Snow                            19
Rain,Snow                                    18
Thunderstorms,Rain Showers                   16
Haze                                         16
Drizzle,Snow,Fog                             15
Freezing Rain                                14
Freezing Drizzle,Snow                        11
Freezing Drizzle      

In [ ]:
# Method 2 - Filtering (exact match only)
df[df['Weather Condition'] == 'Snow'].shape[0]

In [ ]:
# Method 3 - str.contains (catches Snow Showers, Blowing Snow etc.)
df[df['Weather Condition'].str.contains('Snow', case=False, na=False)]['Weather Condition'].value_counts()

---
## Q10. Find all instances when Wind Speed is above 24 AND Visibility is 25

In [ ]:
df[(df['Wind Speed_km/h'] > 24) & (df['Visibility_km'] == 25)]

---
## Q12. What is the Min & Max value of each column against each Weather Condition?

In [ ]:
# Minimum values
df.groupby('Weather Condition').min(numeric_only=True)

In [ ]:
# Maximum values
df.groupby('Weather Condition').max(numeric_only=True)

---
## Q13. Show all the Records where Weather Condition is Fog

In [ ]:
df[df['Weather Condition'] == 'Fog']

---
## Q14. Find all instances when Weather is Clear OR Visibility is above 40

In [ ]:
df[(df['Weather Condition'] == 'Clear') | (df['Visibility_km'] > 40)]

---
## Q15. Find all instances when Weather is Clear AND Relative Humidity > 50

In [ ]:
df[(df['Weather Condition'] == 'Clear') & (df['Rel Hum_%'] > 50)]

---
## ✅ Summary of Key Findings

| Question | Finding |
|---|---|
| Unique Wind Speeds | 34 distinct values (0–83 km/h) |
| Clear Weather | Recorded 1,326 times |
| Wind Speed = 4 km/h | Recorded 474 times |
| Null Values | None — dataset is complete |
| Pressure Std Dev | 0.84 kPa |
| Humidity Variance | 286.25 |
| Snow records (exact) | 390 times |
| Snow records (all types) | 583 times |
| Wind > 24 & Visibility = 25 | 308 records |
| Fog records | 150 records |
| Clear OR Visibility > 40 | 3,027 records |
| Clear AND Humidity > 50 | 1,070 records |